### PR TITLE
Fix transformers v4.42.0 compatibility

### PR DESCRIPTION
**Pull request by echarlaix**
Fixes https://github.com/huggingface/optimum-intel/issues/792

Can be merged after transformers next patch release https://github.com/huggingface/transformers/pull/31755

### CODE REVIEW

File: optimum/intel/neural_compressor/trainer.py [hunk 1]
```diff
@@ -681,6 +681,7 @@ def _inner_training_loop(
     def save_model(
         self,
         output_dir: Optional[str] = None,
+        _internal_call: bool = False,
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/trainer.py:684]: needed for https://github.com/huggingface/transformers/blob/v4.41.0/src/transformers/trainer.py#L2811

**Review by IlyasMoutawwakil — APPROVED** [review]
LGTM Thanks

File: tests/openvino/test_modeling.py [hunk 2]
```diff
@@ -749,6 +748,11 @@ def test_compare_to_transformers(self, model_arch):
         )
 
         ov_outputs = ov_model.generate(**tokens, generation_config=gen_config)
+
+        # TODO: update _update_model_kwargs_for_generation so that it's compatibile with transformers >= v4.42.0
+        if model_arch not in ["chatglm", "glm4"] and is_transformers_version(">=", "4.42.0"):
```
Comment (echarlaix) [inline tests/openvino/test_modeling.py:753]: @eaidova  `_update_model_kwargs_for_generation` needs to be updated to enable generation with transformers v4.42.0 https://huggingface.co/katuni4ka/tiny-random-chatglm2/blob/main/modeling_chatglm.py#L871 as now `_extract_past_from_model_output` also returns `cache_name`

https://github.com/huggingface/transformers/blob/82486e5995ed0a65520b10ce1ea938214a199231/src/transformers/generation/utils.py#L634

Comment (eaidova) [reply tests/openvino/test_modeling.py:753]: ok, I'll update

Comment (echarlaix) [reply tests/openvino/test_modeling.py:753]: thnaks a lot @eaidova

File: setup.py [hunk 3]
```diff
@@ -28,8 +28,9 @@
 
 INSTALL_REQUIRE = [
     "torch>=1.11",
-    "transformers>=4.36.0,<4.42.0",
-    "optimum~=1.20",
+    "transformers>=4.36.0,<4.43.0",
+    "optimum @ git+https://github.com/huggingface/optimum.git",
+    # "optimum>=1.21.2,<1.22.0",
```
Comment (echarlaix) [inline setup.py:33]: needed to include https://github.com/huggingface/optimum/pull/1938

File: tests/openvino/test_modeling.py (hunk 2 above)
Comment (eaidova) [reply tests/openvino/test_modeling.py:753]: https://huggingface.co/katuni4ka/tiny-random-chatglm2/commit/71090c68d20238495ebf43f6c2f9ae194b1a7a2d

**Pull request merged**